test/test_websocket_grid.py

```diff
@@ -0,0 +1,320 @@
+from app.websocket.app import create_app
+from flask import Flask, session, request, json as flask_json
+from flask_socketio import (
+    SocketIO,
+    send,
+    emit,
+    join_room,
+    leave_room,
+    Namespace,
+    disconnect,
+)
+
+import unittest
+import threading
+import multiprocessing
+import grid as gr
+import syft as sy
+import torch as th
+import time
+import requests
+import json
+
+hook = sy.TorchHook(th)
+
+
+gateway_url = "http://localhost:8080"
+
+ports = ["3000", "3001", "3002"]
+ids = ["Bob", "Alice", "James"]
+
+
+def setUpGateway(port):
+    import os
+
+    os.environ["SECRET_KEY"] = "Secretkeyhere"
+    from gateway.app import create_app
+
+    app = create_app(debug=False)
+    app.run(host="0.0.0.0", port="8080")
+
+
+def setUpNodes(config):
+    from app.websocket.app import create_app, socketio
+
+    requests.post(
+        gateway_url + "/join",
+        data=json.dumps(
+            {
+                "node-id": config[0],
+                "node-address": "http://localhost:" + config[1] + "/",
+            }
+        ),
+    )
+    app = create_app(debug=False)
+    socketio.run(app, host="0.0.0.0", port=port)
+
+
+# Init Grid Gateway
+p = multiprocessing.Process(target=setUpGateway, args=("8080",))
+p.daemon = True
+p.start()
+time.sleep(5)
+
+# Init Grid Nodes
+for (node_id, port) in zip(ids, ports):
+    config = (node_id, port)
+    p = multiprocessing.Process(target=setUpNodes, args=(config,))
+    p.daemon = True
+    p.start()
+
+time.sleep(20)
+
+
+class SocketIOAPITest(unittest.TestCase):
+    def setUp(self):
+        self.nodes = []
+        for (node_id, port) in zip(ids, ports):
+            node = gr.WebsocketGridClient(
+                hook, "http://localhost:" + port + "/", id=node_id
+            )
+            node.connect()
+            time.sleep(0.2)
+            self.nodes.append(node)
+
+    def tearDown(self):
+        for node in self.nodes:
+            node.disconnect()
+
+    def test_connect_nodes(self):
+        try:
+            for node in self.nodes:
+                for n in self.nodes:
+                    if n == node:
+                        continue
+                    else:
+                        node.connect_grid_node(n.uri, n.id)
+                        time.sleep(0.2)
+        except:
+            self.fail("test_connect_nodes : Exception raised!")
+
+    def test_multiple_connections_to_same_node(self):
```

Comment:
add a @pytest.mark.parametrize and use a multiple number of connections not only 10. Maybe something like [2, 5, 10] (not sure what are realistic values)